artifact.go

```diff
@@ -215,55 +218,413 @@ func (m artifactModule) postBuildArtifactHandler(c *gin.Context) {
 	if !ok {
 		return
 	}
+	files, ok := parseMultipartFormData(c, buildID)
+	if !ok {
+		return
+	}
+	_, ok = m.createArtifacts(c, files, buildID)
+	if !ok {
+		return
+	}
+	c.Status(http.StatusCreated)
+}
+
+type file struct {
+	name     string
+	fileName string
+	data     []byte
+}
+
+// postTestResultDataHandler godoc
+// @summary Post test result data
+// @tags artifact
+// @accept multipart/form-data
+// @param buildid path int true "Build ID"
+// @param file formData file true "Test result artifact file"
+// @success 201 {object} TestResultListSummary "Added new test result data and created summary"
+// @failure 400 {object} problem.Response "Bad request"
+// @failure 502 {object} problem.Response "Database is unreachable"
+// @router /build/{buildid}/test-result-data [post]
+func (m artifactModule) postTestResultDataHandler(c *gin.Context) {
+	buildID, ok := ginutil.ParseParamUint(c, "buildid")
+	if !ok {
+		return
+	}
+
+	files, ok := parseMultipartFormData(c, buildID)
+	if !ok {
+		return
+	}
+
+	artifacts, ok := m.createArtifacts(c, files, buildID)
+	if !ok {
+		return
+	}
+
+	summaries := make([]TestResultSummary, 0, len(artifacts))
+	lotsOfDetails := make([]TestResultDetail, 0)
+
+	for _, artifact := range artifacts {
+		summary, details, err := getTestSummaryAndDetails(artifact.Data, artifact.ArtifactID, buildID)
+		if err != nil {
+			log.Warn().
+				WithError(err).
+				WithString("filename", artifact.Name).
+				WithUint("build", buildID).
+				WithUint("artifact", artifact.ArtifactID).
+				Message("Failed to unmarshal; invalid TRX/XML format, or the structure" +
+					" receiving the data is malformed.")
```

Comment:
What do you mean with this message "the structure receiving the data is malformed"?

I think it's better to only state what has gone wrong, especially in log messages, and let the reader figure out what the cause it.

Ex:

```suggestion
				Message("Failed to unmarshal TRX/XML test results artifact.")
```